bpf: wireguard: move to cil_from_wireguard program #37084

smagnani96 opened the pull request:
Depends on:
1. [x] https://github.com/cilium/cilium/pull/38179: program detach fix.
2. [x] https://github.com/cilium/cilium/pull/38184: backport of (1).
3. [x] https://github.com/cilium/cilium/pull/38187: land in v1.17 the removal logic for the new `cil_from_wireguard` program we're introducing here, to handle Cilium downgrades.
4. [x] https://github.com/cilium/cilium/pull/38595: land in v1.17 the removal logic for the newly renamed `cilium_calls_wireguard_<ifindex>`

Feature will not be backported.

Notes 1/1 w/ Robin:
* Upgrades from v1.17 to v1.18:
  * TCX: `cil_from_wireguard` is attached to the tail of programs queue. Unless the previous program returns `TCX_NEXT` (and we NEVER use this return code), `cil_from_wireguard` will never see the packet. This means that only `cil_from_netdev` (head of the queue) handles packet processing. Removing `cil_from_netdev` will atomically switch the packet processing to the subsequent program, `cil_from_wireguard`.
  * TC: `cil_from_wireguard` is attached the old-fashioned way, which already atomically attaches the new filter replacing the previous one (`cil_from_netdev` in our case).
* Downgrades from v1.18 to v1.17:
   * TCX: `cil_from_wireguard` would be the head-of-the-tcx-queue program handling traffic. We attach the previous `cil_from_netdev` to the tail. As soon as we remove `cil_from_wireguard`, the `cil_from_netdev` will start handling packet processing atomically.
   * TC: `cil_from_netdev` would be attached the old-fashioned way, instantly replacing the `cil_from_wireguard` program atomically.

Notes 1/1 w/ Timo:
* `MissingTailCall`: when downgrading, the v1.17 bpf_wireguard is compiled, and upon executing the loader command `commit()`, the call map is being overwritten. However, the from_wireguard program introduced in this PR is still running, and it looses the references to the tail call. For this reason, we opt for renaming the calls map for wireguard in v1.18 (this PR), while also landing the removal logic support in v1.17 (linked PR above (4)).

Fixes: #33676

```release-note
Move from bpf_host (from-netdev) to a dedicated program for the WireGuard Ingress hook.
```

rgo3 commented on pkg/datapath/loader/loader.go:
```diff
@@ -256,6 +259,15 @@ func removeObsoleteNetdevPrograms(devices []string) error {
 	ingressDevs := []netlink.Link{}
 	egressDevs := []netlink.Link{}
 	for _, l := range links {
+		// Remove cil_from_netdev program from cilium_wg0 (tcx).
```
While we are removing the from netdev program, I don't think this belongs here. This function was aimed at removing progs from net devices cilium no longer manages, and `cilium_wg0` is certainly a device we still manage after the upgrade. I don't recall the exact ordering but this is potentially called before you attach the `from-wireguard` replacement and thus ends in connectivity disruption.

smagnani96 replied:
Totally agree, thanks for pointing out.
I just moved it after our WireGuard attach (as you suggested below), let's see how it goes in CI.

rgo3 commented on pkg/datapath/loader/loader.go:
```diff
@@ -732,9 +734,27 @@ func replaceWireguardDatapath(ctx context.Context, cArgs []string, device netlin
+		}
+	}
+	if option.Config.NeedIngressOnWireGuardDevice() {
+		if err := attachSKBProgram(device, obj.FromWireguard, symbolFromWireguard,
```
Why can't we detach `from-netdev` once this successfully returns?

smagnani96 replied:
See above reply.

rgo3 reviewed (changes requested):
Quick first pass:

smagnani96 committed:
bpf:refactor: move host_secctx_from_ipcache to global config

From network-facing devices we use this config variable to decide whether
to lookup the secctx from ipcache. Given that we will transition to a
dedicated from-wireguard program in the subsequent commits, it will share
this codepath with bpf_host. So rather than duplicating the code, simply
move the secctx variable into global config: its values is decided in the
agent depending on the legacy routing value, but it is always set to
false for bpf_overlay, bpf_lxc, bpf_network, and bpf_host.

The decision to set this as a global config and not a node config is that
the latter would conceptually set while calling `nodeConfig()` before each
invocation to `NewBPFHost() / NewBPFNetwork` etc. However, given that each
program needs to set a different value for this variable, the global config
is more (logically) suited for these kind of changes.

In BPF tests, we used to assign 1 to this variable to have it != 2, where
2 was the previous value to ensure identity lookups from IPCache are enabled.
Given that host_secctx_from_ipcache is now always false as default, we can
now cleanup all those config assignments, which might be confusing and they
are not needed anymore.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: rename cilium_calls map using ifindex

In this commit we change the name of the cilium_calls map for tail calls
in bpf_wireguard. Prior to this, the map was named `cilium_calls_wireguard_2`,
where '2' stands for identity.ReservedIdentityWorld. Given we do not expect
multiple WireGuard interfaces in Cilium in the same node, it is ok to use
such fixed name. However, renaming the calls map has two main benefits:

1. using ifindex rather than identity.ReservedIdentityWorld aligns the
   behavior to what we do for netdev. We could also consider this transition
   for the map in bpf_overlay, but this won't be part of this patch series.
2. when transitioning to a dedicated from-wireguard program (subsequent
   commits), we preserve connectivity and avoid MissingTailCall errors
   while downgrading to previous versions where we used to attach
   from-netdev to the ingress hook of cilium_wg0. We observed that,
   during downgrades for instance from v1.18 next commit (from-wireguard)
   to v1.17 (from-netdev), the `replaceWireguardDatapath()` function
   clears the cilium_calls map for the WireGuard programs as soon as we
   invoke the `commit()` function. This means that while we wait for the
   from-netdev (v1.17) to be attached, the from-wireguard (v1.18) is running
   with an empty tail call map, since it is being overwritten.
   We cannot first detach from-wireguard, as north-south connections that
   require rev-xlation would be disrupted (pod-to-pod would work through
   the stack rather than us calling bpf_redirect).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard:bpf:feature: move from cil_from_netdev to cil_from_wireguard

We currently need to attach `cil_from_netdev` to the ingress hook of the
Wireguard `cilium_wg0` device to enable encryption of KPR traffic in native
routing. In that case, a reply packet would be otherwise dropped due
to lacking of rev-SNAT/DNAT (for further info, please refer to the Issue/PR
https://github.com/cilium/cilium/pull/19401).

However, we don't need the whole logic, which might be problematic as described
in https://github.com/cilium/cilium/pull/19401#discussion_r1066070869:
in case of host firewall enabled, policy enforcement would not only
(erroneously) happen, but they would also be applied ONLY to the ingress
hook of the wireguard device.

Let's switch to a dedicated program with the minimal set of needed
features. The new `cil_from_wireguard` must look as much as possible to
the previous `cil_from_netdev`, while additionally dropping not needed
features. For this reason, here's the list of codepaths that have been
either dropped or preserved:

* Codepaths dropped:
    * ENABLE_NODEPORT_ACCELERATION
    * ctx->vlan_present
    * ENABLE_HOST_FIREWALL
    * ENABLE_IPSEC
    * ENABLE_ARP_PASSTHROUGH || ENABLE_ARP_RESPONDER || ENABLE_L2_ANNOUNCEMENTS
* Codepaths preserved:
    * ENABLE_IPV4_FRAGMENTS
    * ENABLE_NODEPORT
    * ENABLE_HOST_ROUTING

This means that the new wireguard program, as well as the previously
attached from_netdev, is able to locally deliver packets to endpoints.
However, differently than `cil_from_overlay`, if an endpoint is not
identified, the packet is left for the stack.

This patch supports Cilium version upgrades with the removal of
the previously used cil_from_netdev. Having both the old from-netdev and
the new from-wireguard programs attached causes connectivity problems,
therefore here we make sure to always detach the old program.
For downgrades, we previously landed in v1.17 (i) the removal logic for
the newly introduced calls_map, and (ii) the removal logic for the new
from-wireguard program https://github.com/cilium/cilium/pull/38187: this
means that, upon downgrading Cilium, we are able to:

* TCX: (1) re-attach from-netdev then (2) remove from-wireguard
* TC: replace existing filter with the previous from-netdev

Upgrades/Downgrades are correctly handles also due to changin the cilium
tailcall map used in the wireguard code path in v1.18: if we preserved the
same name cilium_calls_wireguard_2 than v1.17, when downgrading we could
experiment MissingTailCall erorr, given that the new from-wireguard program
would be still in place while the calls map is completely overwritten from
replaceWireguard. The from-wireguard program is later detached upon
re-inserting the old from-netdev, but for that time window it would not
be able to handle packets, leaving them for the stack (in most of the
cases they'd be still reaching their endpoints).

An additional thing to note, is that with the cil_from_wireguard program
inserted we preserve the same packets' path as before when we used to
attach from_netdev rather than from_wg:

* Request pod-to-pod: to_pod->to_ndetdev->to_wg->to_ndetdev
* Reply pod-to-pod:   from_netdev->from_wg->from_pod

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard:bpf:misc: cleanups WG_IFINDEX usages for bpf_host

This patch simplifies the logic where we previously needed to check
whether the program was bpf_wireguard or bpf_host used as ingress for the
WireGuard ingress hook. With the previous commit we moved to a complete
standalone from-wireguard program, therefore we can cleanup codepaths
in which we previously used to check THIS_INTERFACE_IFINDEX == WG_IFINDEX.

In particular, the observation points TRACE_{FROM,TO}_CRYPTO can now be
used with just IS_BPF_WIREGUARD.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard:bpf:tests: refactor and reuse l3_nodeport bpf tests

This commit refactores tc_nodeport_l3_dev.c tests to be able to reuse as
much code as possible between pure `tc_nodeport_l3_dev` (generic L3 device,
not specifically set to wireguard) and `tc_nodeport_l3_wireguard`
(specific to wireguard, where different BPF metrics must be ensured).
Let's keep them both in case of L3 devices not necessarily tighted to
crypto functions (ex. wireguard).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

thorn3r reviewed (approved):
agent changes look good :+1:

rgo3 reviewed (approved):
LGTM, nice work Simone! :rocket:

tommyp1ckles commented on pkg/datapath/loader/loader.go:
```diff
@@ -790,16 +777,31 @@ func replaceWireguardDatapath(ctx context.Context, logger *slog.Logger, lnc *dat
+		)
+	}
+	// Cleanup previous calls map from v1.17.
+	os.RemoveAll(filepath.Join(bpf.TCGlobalsPath(), fmt.Sprintf("cilium_calls_wireguard_%d", identity.ReservedIdentityWorld)))
```
nit: might be best to emit errors that aren't "file not found".

smagnani96 replied:
Since removing such map is not a functional problem, I wouldn't emit errors at this point during the datapath regeneration. Maybe a warning/info. WDYT?

tommyp1ckles replied:
yeah warning makes sense

smagnani96 replied:
I'm not fully convinced yet, will send a future patch in case where I update all the `cleanCallsMaps()` usages 😞 
The error seems not to be checked in most of the `os.Remove` usages, and I'm not sure a warn/info message here  would improve debuggability (the map is not in use anyway).

tommyp1ckles commented on pkg/datapath/loader/loader.go:
```diff
@@ -790,16 +777,31 @@ func replaceWireguardDatapath(ctx context.Context, logger *slog.Logger, lnc *dat
 			)
 		}
 	}
+	// Cleanup previous cil_from_netdev from v1.17.
```
So the implication is we can remove this post 1.19? Maybe add a reminder?

smagnani96 replied:
Yes, in v1.19 (or even v1.18.1) we wouldn't need this anymore, since we've supposedly already taken care of that.

Do you mean a reminder in the code? (if yes, I'd wait in case of other needed changes to restart the CI process)

smagnani96 replied:
Updated comment in the code.

tommyp1ckles reviewed (approved):
cli changes look good

brb reviewed (approved):
Amazing work! Thanks for the detailed commit messages.

smagnani96 committed:
bpf:tests: add egress tests for tc_nodeport_l3_dev.h

This commit completes tests in tc_nodeport_l3_dev.h adding the cases for
egress hooks, both used for IS_BPF_WIREGUARD and IS_BPF_HOST. While adding
these checks, we also ensures that metrics are correctly updated while
processing packets from the Wireguard device with respect to any other
L3 device.

This patch allows us to remove the ad-hoc egress tests previously introduced
in the `wireguard_metrics.c` file, now covered from `tc_nodeport_l3_wireguard.c`.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
